skema/skema-rs/mathml/src/expression.rs

```diff
@@ -14,42 +23,1094 @@ enum Atom {
+                                } else {
+                                    graph.add_edge(
+                                        node_idx,
+                                        parent_node_index,
+                                        op_copy[i].to_string(),
+                                    );
+                                }
+                            }
+                            args[i].to_graph(graph);
+                        }
+                    }
+                }
+            }
+        }
+    }
+}
+
+pub fn redundant_paren(str: &String) -> bool {
+    let str_len = get_str_len(str);
+    if str.chars().nth(0) != Some('(') || str.chars().nth(str_len - 1) != Some(')') {
+        return false;
+    }
+    let mut par_stack = VecDeque::new();
+    par_stack.push_back("left_par");
+    for i in 1..=str_len - 2 {
+        if str.chars().nth(i) == Some('(') {
+            par_stack.push_back("par");
+        } else if str.chars().nth(i) == Some(')') {
+            par_stack.pop_back();
+        }
+    }
+    if par_stack.len() > 0 {
+        if par_stack[0] == "left_par" {
+            return true;
+        }
+    }
+    return false;
+}
+
+pub fn all_multi_div(op: &mut Vec<Operator>) -> bool {
+    for o in 1..=op.len() - 1 {
+        if op[o] != Operator::Multiply && op[o] != Operator::Divide {
+            return false;
+        }
+    }
+    return true;
+}
+
+impl PreExp {
+    fn group_expr(&mut self) {
+        for arg in &mut self.args {
+            match arg {
+                Expr::Atom(_) => {}
+                Expr::Expression { .. } => {
+                    arg.group_expr();
+                }
+            }
+        }
+    }
+
+    fn get_names(&mut self) {
+        for mut arg in &mut self.args {
+            match &mut arg {
+                Expr::Atom(_) => {}
+                Expr::Expression { .. } => {
+                    arg.get_names();
+                }
+            }
+        }
+    }
+
+    fn to_graph(&mut self) -> MathMLGraph {
+        let mut g = MathMLGraph::new();
+        // let root_index = g.add_node("root".to_string());
+        for mut arg in &mut self.args {
+            match &mut arg {
+                Expr::Atom(_) => {}
+                Expr::Expression { .. } => {
+                    arg.to_graph(&mut g);
+                }
+            }
+        }
+        return g;
+    }
+}
+
+pub fn get_str_len(str: &str) -> usize {
+    let mut count = 0;
+    if str.len() > 0 {
+        for i in 0..=str.len() - 1 {
+            match str.chars().nth(i) {
+                None => {}
+                Some(_) => {
+                    count = count + 1;
+                }
+            }
+        }
+    }
+    return count;
+}
+
+pub fn remove_paren(str: &mut String) -> &mut String {
+    while redundant_paren(str) {
+        str.remove(str.len() - 1);
+        str.remove(0);
+    }
+    return str;
+}
+
+// if exists, return the node index; if no, add a new node and return the node index
```

Comment:
Can you change the double slash `//` to a triple slash `///` - that will make it a doc comment: https://doc.rust-lang.org/beta/rust-by-example/meta/doc.html

Reply:
Sure!